src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -136,38 +142,20 @@ protected override async Task OnRun(CancellationToken token)
                         }
                     }
 
-                    DateTime currentTimestamp = _context.Clock.UtcNow.UtcDateTime;
-
-                    // If rule has an action count window, Remove all execution timestamps that fall outside the window.
-                    if (actionCountWindowDuration.HasValue)
-                    {
-                        DateTime windowStartTimestamp = currentTimestamp - actionCountWindowDuration.Value;
-                        while (executionTimestamps.Count > 0)
-                        {
-                            DateTime executionTimestamp = executionTimestamps.Peek();
-                            if (executionTimestamp < windowStartTimestamp)
-                            {
-                                executionTimestamps.Dequeue();
-                            }
-                            else
-                            {
-                                // Stop clearing out previous executions
-                                break;
-                            }
-                        }
-                    }
+                    DateTime currentTimestamp = Context.Clock.UtcNow.UtcDateTime;
 
-                    // Check if executing actions has been throttled due to count limit
-                    if (actionCountLimit > executionTimestamps.Count)
+                    if (!_stateHolder.CheckForThrottling(currentTimestamp))
                     {
-                        executionTimestamps.Enqueue(currentTimestamp);
+                        _stateHolder.AddTimestamp(currentTimestamp);
```

Comment:
Ideally, this is part of CheckForThrottling so that way there is no race condition between checking and adding. I think in another comment you mentioned possibly renaming this method so it's clear that this is doing more than just checking. Maybe `bool BeginActionExecution(DateTime)` then you can fold the existing BeginActionExecution into the previously named CheckForThrottling.

Reply:
The only thing that's a bit tricky is that we also call `CheckForThrottling` when we get a copy of the state, where we aren't adding an action execution. I ended up doing something that I think is a bit different than what you suggested, but is in the same spirit - you can let me know if you want me to adjust it.